src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
After making the above, I decided I'd make absolutely sure that I'd at least considered other alternatives. 

First, I did some more in depth research into Emscripten's LLVM, specifically it's JSBackend. I've found that JSBackend does rewrite many math functions into their JS `Math.*` equivalents. However, it only detects calls to functions from `libm`; Emscripten doesn't expect use of the LLVM math intrinsics, which is supported by the JSBackend requirement on incoming IR: the IR is PNaCl IR simplified.

Second, the following are based on a discussion I had with JF Bastien from Google's NaCl team. The two best options we considered are:

1. Modify the LLVM gold plugin; and
2. Insert an extra `*.ll` file into PNaCl modules before being sent to the linker.

Number 1 means LLVM would, upon loading a bitcode module, look at the intrinsic declarations present and add the `libm` equivalents to the undefined symbols reported back to the linker. Later, a PNaCl IR pass would do the actual rewrite to the `libm` function. This is ugly, because this would force the gold plugin to treat PNaCl targets differently vs the general case (and Rust modules don't actually have their target triple set to the PNaCl target triple!).

Number 2 avoids all that, but sadly isn't really any cleaner (in fact, I'd say it's worse): it involves making the LLVM intrinsics alias the `libm` versions. Apparently, this isn't *completely* undefined behavior.


I still think doing the redirection in Rust is the better route.